app/lib/search/index_simple.dart

```diff
@@ -0,0 +1,208 @@
+        list = [];
+      } else {
+        list = list.sublist(query.offset);
+      }
+    }
+    if (query.limit != null && list.length > query.limit) {
+      list = list.sublist(0, query.limit);
+    }
+
+    return new PackageSearchResult(
+      totalCount: totalCount,
+      indexUpdated: _lastUpdated.toIso8601String(),
+      packages: list.map((pr) {
+        final PackageDocument doc = _documents[pr.url];
+        return new PackageScore(
+          url: doc.url,
+          package: doc.package,
+          version: doc.version,
+          devVersion: doc.devVersion,
+          score: pr.score,
+        );
+      }).toList(),
+    );
+  }
+
+  @override
+  Future merge() async {
+    _isReady = true;
+    _lastUpdated = new DateTime.now().toUtc();
+  }
+
+  @override
+  Future<int> indexSize() async => -1;
+}
+
+class _TokenIndex {
+  final Map<String, Set<String>> _inverseUrls = {};
+  final Map<String, double> _weights = {};
+
+  void add(String url, String text) {
+    final Set<String> tokens = _tokenize(normalizeBeforeIndexing(text));
+    if (tokens == null || tokens.isEmpty) return;
+    double sumWeight = 0.0;
+    for (String token in tokens) {
+      final Set<String> set = _inverseUrls.putIfAbsent(token, () => new Set());
+      set.add(url);
+      sumWeight += _tokenWeight(token);
+    }
+    _weights[url] = sumWeight;
+  }
+
+  void removeUrl(String url) {
+    _weights.remove(url);
+    for (Set<String> set in _inverseUrls.values) {
+      set.remove(url);
+      // TODO: remove set if it becomes empty
+    }
+  }
+
+  // A TF-IDF-like scoring, with more weight for longer terms.
+  Map<String, double> search(String text) {
+    final Set<String> tokens = _tokenize(normalizeBeforeIndexing(text));
+    if (tokens == null || tokens.isEmpty) return null;
+    double sumWeight = 0.0;
+    final Map<String, double> counts = {};
+    for (String token in tokens) {
+      final Set<String> set = _inverseUrls[token];
+      if (set == null || set.isEmpty) continue;
+
+      final double weight = _tokenWeight(token) / set.length;
+      sumWeight += weight;
+
+      for (String url in set) {
+        final double prevValue = counts[url] ?? 0.0;
+        counts[url] = prevValue + weight;
+      }
+    }
+    for (String url in counts.keys.toList()) {
+      counts[url] = 10000.0 * counts[url] / sumWeight / _weights[url];
+    }
+    return counts;
+  }
+
+  double _tokenWeight(String token) => (token.length * token.length).toDouble();
+}
+
+Set<String> _tokenize(String text) {
+  text = normalizeBeforeIndexing(text);
+  if (text.isEmpty) return null;
+  final Set<String> ngrams = new Set();
+  for (int ngramLength = 1; ngramLength <= 4; ngramLength++) {
+    if (text.length <= ngramLength) {
+      ngrams.add(text);
+    } else {
+      for (int i = 0; i <= text.length - ngramLength; i++) {
+        ngrams.add(text.substring(i, i + ngramLength));
+      }
+    }
+  }
+  return ngrams;
+}
+
+class _PackageResult {
+  String url;
```

Comment:
made `url` final, but `score` is more complicated, mostly because of the way I update it with popularity score. To rewrite that part (and align it with the other weights) requires a bit more iterations, I'll happy to do that in a subsequent change too.